game-core/src/test/java/games/strategy/triplea/ai/tree/BattleStepTest.java

```diff
@@ -0,0 +1,940 @@
+    final BattleStep root = new BattleStep(attackingUnits, BRITISH, 0, createParameters());
+    return root.getRegularDiceGrouped(parameters, false, attackers, defenders);
+  }
+
+  @Test
+  void getDiceGroupedWithSupport() {
+    final List<Unit> attackingOrderOfLoss = new ArrayList<>();
+    attackingOrderOfLoss.add(new Unit(INFANTRY, BRITISH, data));
+    attackingOrderOfLoss.add(new Unit(ARTILLERY, BRITISH, data));
+
+    final Map<Integer, Integer> diceGrouped = getDiceGrouped(attackingOrderOfLoss, List.of());
+
+    final Map<Integer, Integer> expected = new HashMap<>();
+    expected.put(2, 2);
+
+    assertEquals(expected, diceGrouped);
+  }
+
+  @Test
+  void getDiceGroupedWithSupportExtraNonSupport() {
+    final List<Unit> attackingOrderOfLoss = new ArrayList<>();
+    attackingOrderOfLoss.add(new Unit(INFANTRY, BRITISH, data));
+    attackingOrderOfLoss.add(new Unit(INFANTRY, BRITISH, data));
+    attackingOrderOfLoss.add(new Unit(ARTILLERY, BRITISH, data));
+
+    final Map<Integer, Integer> diceGrouped = getDiceGrouped(attackingOrderOfLoss, List.of());
+
+    final Map<Integer, Integer> expected = new HashMap<>();
+    expected.put(2, 2);
+    expected.put(1, 1);
+
+    assertEquals(expected, diceGrouped);
+  }
+
+  @Test
+  void calculateSingleHitProbability() {
+    final List<Unit> attackingOrderOfLoss = new ArrayList<>();
+    attackingOrderOfLoss.add(new Unit(FIGHTER, BRITISH, data));
+
+    final int diceSides = data.getDiceSides();
+    final Map<Integer, Integer> diceGrouped = getDiceGrouped(attackingOrderOfLoss, List.of());
+    final StepUnits attackingUnits =
+        new StepUnits(attackingOrderOfLoss, BRITISH, List.of(), GERMAN);
+    final BattleStep root = new BattleStep(attackingUnits, BRITISH, 0, createParameters());
+    final List<Double> hitProbability =
+        root.calculateHitProbabilities(BattleStep.RollData.of(1, diceGrouped, diceSides));
+
+    assertEquals(0.5, hitProbability.get(0));
+    assertEquals(0.5, hitProbability.get(1));
+  }
+
+  @Test
+  void calculateDoubleHitProbability() {
+    final List<Unit> attackingOrderOfLoss = new ArrayList<>();
+    attackingOrderOfLoss.add(new Unit(INFANTRY, BRITISH, data));
+    attackingOrderOfLoss.add(new Unit(ARMOUR, BRITISH, data));
+
+    final int diceSides = data.getDiceSides();
+    final Map<Integer, Integer> diceGrouped = getDiceGrouped(attackingOrderOfLoss, List.of());
+    final StepUnits attackingUnits =
+        new StepUnits(attackingOrderOfLoss, BRITISH, List.of(), GERMAN);
+    final BattleStep root = new BattleStep(attackingUnits, BRITISH, 0, createParameters());
+    final List<Double> hitProbability =
+        root.calculateHitProbabilities(BattleStep.RollData.of(2, diceGrouped, diceSides));
+
+    assertEquals(0.4166, hitProbability.get(0), 0.0001);
+    assertEquals(0.5, hitProbability.get(1), 0.0001);
+    assertEquals(0.0833, hitProbability.get(2), 0.0001);
```

Comment:
I assume those are the probabilities for win, lose, draw?

If the List was a dedicated type instead, this would be clearer

Reply:
These are probabilities of hits.  The first line is 0 hits (out of 2 dice), the second is 1 hit, the third is 2 hits.  hitProbability can grow depending on the number of dice rolled.